<h1>Q-Learning</h1>

<p>Στην συγκεκριμένη άσκηση θα μελετήσετε το στοχαστικό αλγόριθμο Q-Learning, χρησιμοποιώντας το έτοιμο πρόγραμμα που σας δίνεται.</p> <p> Στα πλαίσια του παραδείγματος θα εξετάσετε μία υλοποίηση με Q-learning σχετικά με το σύστημα αυτόματης οδήγησης ενός ταξί. Στα πλαίσια του προβλήματος αυτού θα πρέπει να ικανοποιούνται τα εξής:</p>
<ul>
<li>Το ταξί θα πρέπει να αφήνει τον πελάτη στη σωστή θέση</li>
<li>Το ταξί να ακολουθεί τη συντομότερη δυνατή διαδρομή</li>
<li>Να τηρούνται οι κανόνες κυκλοφορίας και ασφάλειας των επιβατών</li>
</ul>

<p>Στα πλαίσια του προβλήματος έχουμε τα εξής χαρακτηριστικά για την επιβράβευση, τις καταστάσεις και τις ενέργειες.</p> 

<h4>Επιβράβευση</h4>
<ul>
<li>Θα έχουμε τη μέγιστη επιβράβευση όταν το ταξί αφήνει έναν πελάτη στην επιθυμητή θέση</li>
<li>Θα υπάρχει ποινή στην περίπτωση όπου το ταξί αφήσει τον πελάτη σε κάποιο λανθασμένο σημείο</li>
<li>Ο agent θα παίρνει μία μικρή σχετικά ποινή στην περίπτωση όπου αργεί να φτάσει στον τελικό προορισμό</li>
</ul>

<p>Γενικά οι παραπάνω αρχές συνοψίζονται στα εξής: "Λαμβάνουμε +20 πόντους για μια επιτυχημένη πτώση και χάνουμε 1 πόντο για κάθε χρονικό βήμα που παίρνει. Υπάρχει επίσης ποινή 10 πόντων για παράνομες ενέργειες παραλαβής και αποχώρησης."</p>

<h4>Πλήθος Καταστάσεων</h4>
<img src="https://storage.googleapis.com/lds-media/images/Reinforcement_Learning_Taxi_Env.width-1200.png">
<p>Στο παράδειγμά μας έχουμε ένα μικρό χώρο 5*5. Από εκεί και πέρα έχουμε 4 προορισμούς και 5 πιθανές θέσεις του πελάτη (παίρνουμε και την περίπτωση ο πελάτης να είναι ήδη μέσα στο τάξι).</p>
<p>Με βάση τα παραπάνω έχουμε 5*5*5*4=500 πιθανές καταστάσεις.</p>

<h4>Πλήθος Ενεργειών</h4>
<p>Έχουμε έξι ενεργειες για το ταξί, οι οποίες είναι οι εξής:</p>
<ul>
<li>0=Νότια</li>
<li>1=Βόρεια</li>
<li>2=Ανατολικά</li>
<li>3=Δυτικά</li>
<li>4=Επιβίβαση</li>
<li>5=Αποβίβαση</li>
</ul>





<p>Πριν συνεχίσετε στην άσκηση να απαντήσετε στο εξής ερώτημα:</p>

<b><p>1. Να περιγράψετε σύντομα τον αλγόριθμο Q-Learning. Σε ποια προβλήματα θεωρείτε ότι ταιριάζει ως τρόπος εκμάθησης η Ενισχυτική Μάθηση (Reinforcement Learning); Ποια είναι η βασική διαφορά του αλγορίθμου Q-Learning από τους αλγορίθμους Policy Iteration και Value Iteration;</p></b>

<font color='#486393'>Σημειώνεται πως οι απαντήσεις στα ερωτήματα γράφονται σε αυτό το χρώμα, προκειμένου να διαφοροποιούνται από την εκφώνηση και τις οδηγίες της άσκησης.</font>

<font color='#486393'>Ο αλγόριθμος Q-learning αντιστοιχεί σε έναν αλγόριθμο ενισχυτικής μάθησης ο οποίος προσπαθεί να προσεγγίσει την τιμή της βέλτιστης value function κατά Bellman, χωρίς να υπολογίζει παράλληλα κάποια βέλτιστη πολιτική, αφού δεν προϋποθέτει τη μοντελοποίηση του περιβάλλοντος του εκάστοτε προβλήματος. Προκειμένου να αποτυπωθεί μαθηματικά η αρχή λειτουργίας του αλγορίθμου, συμβολίζουμε (ακολουθώντας το φορμαλισμό των Sutton & Barto) την κατάσταση στην οποία βρίσκεται ο πράκτορας (agent) σε κάποιο βήμα, $t$, ως $S_t$, την προτεινόμενη κίνηση για το επόμενο βήμα ως $A_{t+1}$ και την πιθανή επιβράβευση λόγω της κίνησης αυτής ως $R_{t+1}$. Βάσει αυτών, ο αλγόριθμος Q-learning προσπαθεί να υπολογίσει την τιμή της βέλτιστης value function κατά Bellman μέσω της επαναληπτικής διαδικασίας</font>

<font color='#486393'>$$ Q^\text{new}\left(S_t,A_{t+1}\right) = Q^\text{old}\left(S_t,A_{t+1}\right) + \alpha \cdot \left\{R_{t+1} + \gamma\left[\text{max}_{A_{t+1}}{Q^\text{old}\left(S_{t+1},A_{t+1}\right)}\right] - Q^\text{old}\left(S_t,A_{t+1}\right) \right\}, $$</font>

<font color='#486393'>όπου $\alpha$ είναι ο ρυθμός εκμάθησης της διαδικασίας και $\gamma$ είναι ο ρυθμός έκπτωσης (discount factor). Όπως αναλύθηκε και στο πρώτο notebook της παρούσας εργαστηριακής αναφοράς, ο ρυθμός έκπτωσης επιτελεί την εξής λειτουργία: εάν ο πράκτορας λάβει κάποια επιβράβευση $R_{t+m}$ στο βήμα $t+m$ αντί για το βήμα $t$, τότε αυτή θα συνδέεται με την επιβράβευση, $R_{t+1}$, εάν αυτή είχε ληφθεί στο βήμα $t$, μέσω της σχέσης $R_{t+m} = \gamma^{m-1}R_{t+1}$.</font>

<font color='#486393'>Σε γενικές γραμμές, η ενισχυτική μάθηση χρησιμοποιείται σε προβλήματα τα οποία μπορούν να αποτυπωθούν ως Markov Decision Processes (MDPs), προκειμένου να αξιοποιηθούν τεχνικές του δυναμικού προγραμματισμού. Επιπλέον, παρότι και οι αλγόριθμοι της ενισχυτικής μάθησης συχνά αντιστοιχούν σε προβλήματα βελτιστοποίησης, η βελτιστοποίηση αυτή δεν νοείται με τον ίδιο τρόπο που νοείται σε προβλήματα εποπτευόμενης μάθησης, όπου το τελικό output πρέπει να αποκλίνει όσο το δυνατό λιγότερο από ένα συγκεκριμένο target value. Αντίθετα, η ενισχυτική μάθηση χρησιμοποιείται όταν κρίνεται αναγκαία η εύρεση μιας λεπτής ισορροπίας ανάμεσα στην αξιοποίηση σε υψηλό βαθμό της ήδη γνωστής πληροφορίας για το σύστημα και την εξερεύνηση νέων στρατηγικών. Οι νέες αυτές στρατηγικές σε πολλές περιπτώσεις ενδέχεται να οδηγούν και σε χαμηλότερη απόδοση σε πρώτη φάση, όμως η εξερεύνησή τους πρέπει να πραγματοποιείται, ώστε να συγκεντρώνεται νέα πληροφορία η οποία δύναται μακροπρόθεσμα να οδηγήσει σε αύξηση της απόδοσης.</font>

<font color='#486393'>Τέλος, σε ό,τι αφορά την κύρια διαφορά του αλγορίθμου Q-learning από τους αλγορίθμους Policy και Value Iteration που αναλύθηκαν στο πρώτο notebook της παρούσας εργαστηριακής άσκησης, αυτή αναφέρθηκε και παραπάνω: ο αλγόριθμος Q-learning δεν πραγματοποιεί υποθέσεις για το περιβάλλον του υπό μελέτη συστήματος. Είναι, δηλαδή, ένας model-independent αλγόριθμος, γεγονός το οποίο συμβάλλει στην καθολικότητά του (δηλαδή τη χρήση του σε πληθώρα προβλημάτων) και κυρίως στη μείωση της χρονικής του πολυπλοκότητας. Η μείωση αυτή είναι συνέπεια του γεγονότος πως ο αλγόριθμος Q-learning είναι off-policy, δηλαδή δεν ασχολείται με την αρχικοποίηση ή/και βελτιστοποίηση κάποιας πολιτικής.</font>

<p>Στη συνέχεια θα πρέπει να φορτώσετε τη βιβλιοθήκη gym καθώς και το σχετικό dataset<p>

In [1]:
import gym

env = gym.make("Taxi-v3").env

env.render()

+---------+
|R: | : :G|
| : | : : |
| : : : : |
| | : | : |
|Y| : |B: |
+---------+



In [2]:
env.reset() # reset environment to a new, random state
env.render()

print("Action Space {}".format(env.action_space))
print("State Space {}".format(env.observation_space))


+---------+
|R: | : :G|
| : | : : |
| : : : : |
| | : | : |
|Y| : |B: |
+---------+

Action Space Discrete(6)
State Space Discrete(500)


<p>Παρακάτω ορίζουμε τις συνεταγμένες του ταξί, τη θέση του πελάτη και το σημείο προορισμού</p>

In [3]:
state = env.encode(3, 1, 2, 0) # (taxi row, taxi column, passenger index, destination index)
print("State:", state)

env.s = state
env.render()

State: 328
+---------+
|R: | : :G|
| : | : : |
| : : : : |
| | : | : |
|Y| : |B: |
+---------+



<p>Παρακάτω είναι η μήτρα επιβράβευσης για το state που ορίσαμε στο προηγούμενο βήμα</p>

In [4]:
env.P[328]

{0: [(1.0, 428, -1, False)],
 1: [(1.0, 228, -1, False)],
 2: [(1.0, 348, -1, False)],
 3: [(1.0, 328, -1, False)],
 4: [(1.0, 328, -10, False)],
 5: [(1.0, 328, -10, False)]}

<p> Τρέχουμε το παράδειγμά μας χωρις τη χρήση Q-Learning.</p>

<b><p>2. Τα αποτελέσματα είναι ικανοποιητικά; Πως θα μας εξυπηρετούσε η χρήση του Q-Learning;</p></b>

In [5]:
def without_qlearning(episodes=100):
    total_epochs, total_penalties = 0, 0
    episodes = episodes
    rewards = []
    
    for _ in range(episodes):
        env.s = 328  # set environment to illustration's state

        epochs = 0
        penalties, reward = 0, 0

        frames = [] # for animation

        done = False

        while not done:
            action = env.action_space.sample()
            state, reward, done, info = env.step(action)
            rewards.append(reward)

            if reward == -10:
                penalties += 1

            # Put each rendered frame into dict for animation
            frames.append({
                'frame': env.render(mode='ansi'),
                'state': state,
                'action': action,
                'reward': reward
                }
            )

            epochs += 1
            
        total_penalties += penalties
        total_epochs += epochs
        
    if episodes==1:
        print("Timesteps taken: {}".format(epochs))
        print("Penalties incurred: {}".format(penalties))
    else:
        print(f"Results after {episodes} episodes:")
        print(f"Average timesteps per episode: {total_epochs / episodes:.2f}")
        print(f"Average penalties per episode: {total_penalties / episodes:.2f}")
        rewards = np.asarray(rewards)
        print(f"Average reward per step: {rewards.mean():.2f}")

In [6]:
without_qlearning(1)

Timesteps taken: 727
Penalties incurred: 236


<font color='#486393'>Προφανώς, τα αποτελέσματα που προκύπτουν δεν είναι ικανοποιητικά (πάνω από το 30% περίπου των κινήσεων του πράκτορα αντιστοιχούν σε σφάλμα). Ο λόγος για αυτό είναι πως για την επίλυση του προβλήματος δεν ακολουθείται κάποιος αλγόριθμος που να λαμβάνει υπ' όψιν τα σφάλματα και την ανταμοιβή που λαμβάνει ο πράκτορας κατά τη διάρκεια κάθε δρομολογίου. Αυτά, δύνανται να προσφέρουν στον πράκτορα πληροφορία την οποία μπορεί να αξιοποιήσει σε μετέπειτα δρομολόγια για να μεγιστοποιήσει την ανταμοιβή του και παράλληλα να ελαχιστοποιήσει τα σφάλματά του. Με βάση την ανάλυση που προηγήθηκε, ο αλγόριθμος Q-learning αποτελεί έναν τέτοιο αλγόριθμο, ο οποίος θα επιτρέψει στον πράκτορα να βελτιστοποιήσει τα δρομολόγιά του. Η χρήση του έναντι των αλγορίθμων που χρησιμοποιήθηκαν στο άλλο notebook της παρούσας εργαστηριακής άσκησης οφείλεται στο μειωμένο χρόνο εκτέλεσής του (όντας off-policy), χωρίς αυτός να συνεπάγεται αβεβαιότητα ως προς τη σύγκλιση, η οποία είναι εξασφαλισμένη βάσει θεωρημάτων.</font>

<p>Τώρα θα προσπαθήσουμε να λύσουμε το πρόβλημά μας με τη χρήση του Q-Learning.</p>

<b><p>3. Τι γνωρίζετε για τις παραμέτρους α και γ. Τι θα συμβεί αν έχουν τιμές ίσες με 1;</p></b>

<font color='#486393'>Το ερώτημα αυτό αναπτύχθηκε εν μέρει κατά την παρουσίαση του αλγορίθμου Q-learning, παρ' όλα αυτά αξίζει εδώ να επαναληφθούν κάποια βασικά σημεία και να δοθούν περισσότερες λεπτομέρειες σε άλλα.</font>

<font color='#486393'>Σε ό,τι αφορά την παράμετρο $\alpha$, αυτή αντιστοιχεί στο ρυθμό εκμάθησης της επαναληπτικής διαδικασίας. Συγκεκριμένα, η επαναληπτική διαδικασία του αλγορίθμου Q-learning σχετίζεται με την ανανέωση της value function, η οποία προκύπτει ως μια μίξη ανάμεσα στην προηγούμενη τιμή της και τον επιπλέον όρο που σχετίζεται με την ανταμοιβή του πράκτορα στο επόμενο χρονικό βήμα. Στην οριακή περίπτωση όπου $\alpha = 0$ ο πράκτορας δρα χωρίς να μαθαίνει μέσω της αλληλεπίδρασής του με το περιβάλλον (ισοδύναμα με ό,τι παρατηρήσαμε εκτελώντας την προσομοίωση απουσία Q-learning). Στο άλλο όριο, η τιμή $\alpha = 1$ οδηγεί σε εκμάθηση η οποία δεν κρατάει μνήμη και ως εκ τούτου μπορεί να οδηγήσει σε μια καθαρά στοχαστική συμπεριφορά, άνευ στρατηγικής.</font>

<font color='#486393'>Η παράμετρος $\gamma$ αντιστοιχεί στο ρυθμό έκπτωσης, δηλαδή το κατά πόσο η αξία της αναμενόμενης ανταμοιβής που θα λάβει ο πράκτορας μειώνεται σε βάθος χρονικών βημάτων, όπως αναλύθηκε παραπάνω. Εδώ, το όριο $\gamma = 0$ αντιστοιχεί σε έναν «επιπόλαιο» πράκτορα, ο οποίος αποφασίζει καθαρά βάσει της δράσης που θα μεγιστοποιήσει την άμεση ανταμοιβή του. Από την άλλη, το όριο $\gamma = 1$ αντιστοιχεί σε μια κατάσταση όπου με το πέρασμα των χρονικών βημάτων η αναμενόμενη ανταμοιβή του πράκτορα δε μειώνεται. Μάλιστα, όπως αναλύθηκε και στο άλλο notebook, στην περίπτωση όπου $\gamma = 1$ ο αλγόριθμος δε συγκλίνει, καθώς η αναμενόμενη τιμή της value function κατά Bellman αντιστοιχεί σε μια αποκλίνουσα σειρά.</font>

In [7]:
import numpy as np
q_table = np.zeros([env.observation_space.n, env.action_space.n])

In [8]:
%%time
"""Training the agent"""

import random
from IPython.display import clear_output

# Hyperparameters
alpha = 0.1
gamma = 0.6
epsilon = 0.1

# For plotting metrics
all_epochs = []
all_penalties = []

for i in range(1, 100001):
    state = env.reset()

    epochs, penalties, reward, = 0, 0, 0
    done = False
    
    while not done:
        if random.uniform(0, 1) < epsilon:
            action = env.action_space.sample() # Explore action space
        else:
            action = np.argmax(q_table[state]) # Exploit learned values

        next_state, reward, done, info = env.step(action) 
        
        old_value = q_table[state, action]
        next_max = np.max(q_table[next_state])
        
        new_value = (1 - alpha) * old_value + alpha * (reward + gamma * next_max)
        q_table[state, action] = new_value

        if reward == -10:
            penalties += 1

        state = next_state
        epochs += 1
        
    if i % 100 == 0:
        clear_output(wait=True)
        print(f"Episode: {i}")

print("Training finished.\n")

Episode: 100000
Training finished.

Wall time: 29.4 s


In [9]:
q_table[328]

array([ -2.41232915,  -2.27325184,  -2.39641613,  -2.36176221,
       -10.81546216, -10.67772066])

In [10]:
"""Evaluate agent's performance after Q-learning"""
def with_qlearning(episodes=100):
    total_epochs, total_penalties = 0, 0
    episodes = episodes
    rewards = []

    for _ in range(episodes):
        state = env.reset()
        epochs, penalties, reward = 0, 0, 0

        done = False

        while not done:
            action = np.argmax(q_table[state])
            state, reward, done, info = env.step(action)
            rewards.append(reward)

            if reward == -10:
                penalties += 1

            epochs += 1

        total_penalties += penalties
        total_epochs += epochs

    print(f"Results after {episodes} episodes:")
    print(f"Average timesteps per episode: {total_epochs / episodes:.2f}")
    print(f"Average penalties per episode: {total_penalties / episodes:.2f}")
    rewards = np.asarray(rewards)
    print(f"Average reward per step: {rewards.mean():.2f}")

<b><p>4. Συγκρίνετε τους δύο αλγορίθμους με βάση τις παρακάτω μετρικές</p>
<ul>
<li>Μέσος αριθμός παραβάσεων ανά επεισόδιο</li>
<li>Μέσος αριθμός βημάτων ανά διαδρομή</li>
<li>Μέσος αριθμός ανταμοιβών ανά κίνηση</li>
</ul>
<p>Τις παραπάνω συγκρίσεις να τις κάνετε για 100 επεισόδια.</p>
</b>

<font color='#486393'>Έχοντας επεξεργαστεί ελαφρώς το δοθέντα κώδικα (γράφοντάς τον υπό τη μορφή δύο συναρτήσεων), στο ακόλουθο κελί εκτελείται η προσομοίωση 100 επεισοδίων (α) με και (β) χωρίς την εφαρμογή του αλγορίθμου Q-learning.</font>

In [11]:
print('Running simulation for 100 episodes without Q-learning...')
print('*'*57)
without_qlearning(100)
print('\n\n')
print('Running simulation for 100 episodes with Q-learning...')
print('*'*54)
with_qlearning(100)

Running simulation for 100 episodes without Q-learning...
*********************************************************
Results after 100 episodes:
Average timesteps per episode: 2470.93
Average penalties per episode: 803.51
Average reward per step: -3.92



Running simulation for 100 episodes with Q-learning...
******************************************************
Results after 100 episodes:
Average timesteps per episode: 13.03
Average penalties per episode: 0.00
Average reward per step: 0.61


<font color='#486393'>Με βάση τα αποτελέσματα των προσομοιώσεων για 100 επεισόδια, καθίσταται εμφανές πως ο αλγόριθμος Q-learning οδηγεί σε καταπληκτικά αποτελέσματα. Στην περίπτωση των τυχαίων προσομοιώσεων, ο μέσος αριθμός παραβάσεων ανά επεισόδιο είναι περίπου ίσος με 804 παραβάσεις, σε βάθος 2471 βημάτων (κατά μέσο όρο), γεγονός που υποδεικνύει πως το 32.54% των βημάτων κάθε επεισοδίου αντιστοιχεί σε σφάλμα. Από την άλλη, ο εκπαιδευμένος πράκτορας δεν πραγματοποιεί κατά μέσο όρο κανένα σφάλμα ανά επεισόδιο, σε βάθος μόλις 13 περίπου βημάτων κατά μέσο όρο. Τα αποτελέσματα αυτά δε δείχνουν μόνο πως ο πράκτορας πραγματοποιεί σφάλματα στο 0% των βημάτων κάθε επεισοδίου, αλλά και ότι έχει βελτιστοποιήσει τη στρατηγική του ώστε να ολοκληρώνει κάθε επεισόδιο σε μόλις 13 βήματα, δηλαδή δύο τάξεις μεγέθους λιγότερα απ' ότι ο τυχαίος αλγόριθμος. Τέλος, η βελτιστοποίηση αυτή αποτυπώνεται και μέσω της μέσης ανταμοιβής ανά βήμα, η οποία ισούται με -3.92 στην περίπτωση του τυχαίου αλγορίθμου (δηλαδή συνολική ζημία) και με 0.61 στην περίπτωση του εκπαιδευμένου πράκτορα (δηλαδή συνολικό κέρδος).</font>